lib/PipelineC/PipelineCTracing.cpp

```diff
@@ -0,0 +1,283 @@
+
+// The opposite of a std::recursive_mutex, if locked by the same thread it will
+// assert (this is to avoid a deadlock/malformed output when tracing)
+class NonRecursiveMutex {
+private:
+  std::optional<std::thread::id> ThreadId;
+  std::mutex TheLock;
+
+public:
+  NonRecursiveMutex() {}
+
+  void lock() {
+    if (ThreadId.has_value())
+      revng_assert(std::this_thread::get_id() != *ThreadId,
+                   "NonRecursiveMutex entered twice by the same thread!");
+    TheLock.lock();
+    ThreadId = std::this_thread::get_id();
+  }
+
+  void unlock() {
+    revng_assert(std::this_thread::get_id() == *ThreadId);
+    TheLock.unlock();
+    ThreadId.reset();
+  }
+};
+
+static NonRecursiveMutex TraceLock;
+
+// Helper class for tracing, this will be used by the the argument/return
+// handlers defined below to properly print the value of the arguments onto the
+// YAML tracing file
+class PipelineCTracer {
+private:
+  std::string Path;
+  std::error_code OpenEC;
+  llvm::raw_fd_ostream OS;
+  bool OutputtingArguments;
+
+public:
+  PipelineCTracer(std::string Path) :
+    Path(Path), OpenEC(), OS(Path, OpenEC), OutputtingArguments(false) {
+    revng_assert(!OpenEC);
+    printHeader();
+  }
+
+  ~PipelineCTracer() {
+    OS.flush();
+    OS.close();
+  }
+
+private:
+  void printHeader() {
+    OS << "version: 1\n";
+    OS << "commands:\n";
+    OS.flush();
+  }
+
+  std::string reprString(const char *String) {
+    return '"' + llvm::StringRef(String).str() + '"';
+  }
+
+public:
+  void functionPrelude(const llvm::StringRef Name) {
+    OS << "- name: " << Name << "\n";
+    OS << "  arguments:\n";
+    OutputtingArguments = true;
+    OS.flush();
+  }
+
+  void newArgument() {
+    OS << "  - ";
+    OS.flush();
+  }
+
+  template<typename T>
+  // clang-format off
+  requires std::is_same_v<T, uint8_t> ||
+           std::is_same_v<T, uint32_t> ||
+           std::is_same_v<T, uint64_t>
+  // clang-format on
+  void printValue(const T &Int) {
+    OS << Int << "\n";
+    OS.flush();
+  }
+
+  template<typename T>
+  // clang-format off
+  requires std::is_same_v<T, char *> ||
+           std::is_same_v<T, const char *>
+  // clang-format on
+  void printValue(T String) {
+    if (OutputtingArguments)
+      OS << reprString(String);
+    else
+      OS << "P" << (void *) String;
+    OS << "\n";
+    OS.flush();
+  }
+
+  template<typename T>
+  requires std::is_same_v<T, bool>
+  void printValue(const bool &Bool) {
+    OS << (Bool ? "true" : "false") << "\n";
+    OS.flush();
+  }
+
+  template<typename T>
+  // clang-format off
+  requires (std::is_pointer_v<T> && isRPType<remove_constptr<T>>())
+  void printValue(T Ptr) {
+    // clang-format on
+    OS << "P" << (void *) Ptr << "\n";
+    OS.flush();
+  }
+
+  void printPointer(void *Ptr) {
+    // clang-format on
+    OS << "P" << Ptr << "\n";
+    OS.flush();
+  }
+
+  void printBuffer(const char *Buffer, const uint64_t Length) {
```

Comment:
Can we take an `ArrayRef` instead?

Reply:
If anything, it should be `llvm::StringRef` since `encodeBase64` accepts that

Reply:
Works, the goal is to reduce C-isms to only the C API surface.

Reply:
Done